packages/astro/src/core/errors/errors-data.ts

```diff
@@ -628,6 +628,94 @@ See https://docs.astro.build/en/guides/server-side-rendering/ for more informati
 		code: 3030,
 		message: 'The response has already been sent to the browser and cannot be altered.',
 	},
+
+	/**
+	 * @docs
+	 * @description
+	 * Thrown when a middleware doesn't return anything or doesn't call the `next` function.
+	 *
+	 * For example:
+	 * ```ts
+	 * import {defineMiddleware} from "astro/middleware";
+	 * export const onRequest = defineMiddleware((context, _) => {
+	 * 	// doesn't return anything or doesn't call next
+	 * 	context.locals.someData = false;
+	 * });
+	 * ```
+	 */
+	MiddlewareNoDataOrNextCalled: {
+		title: "The middleware didn't return a response or called `next`",
+		code: 3031,
+		message: 'The middleware needs to return a `Response` object or call the `next` function.',
+	},
+
+	/**
+	 * @docs
+	 * @description
+	 * Thrown in development mode, when a middleware return something that is not a `Response`
+	 *
+	 * For example:
+	 * ```ts
+	 * import {defineMiddleware} from "astro/middleware";
+	 * export const onRequest = defineMiddleware(() => {
+	 *   return "string"
+	 * });
+	 * ```
+	 */
+	MiddlewareNotAResponse: {
+		title: 'The middleware returned something that is not a `Response`',
+		code: 3032,
+		message:
+			'When returning something from a middleware, you must return a valid `Response`. Failing doing so would cause errors..',
+	},
```

Comment:
```suggestion
	 * Thrown in development mode when middleware returns something that is not a `Response` object.
	 *
	 * For example:
	 * ```ts
	 * import {defineMiddleware} from "astro/middleware";
	 * export const onRequest = defineMiddleware(() => {
	 *   return "string"
	 * });
	 * ```
	 */
	MiddlewareNotAResponse: {
		title: 'The middleware returned something that is not a `Response` object',
		code: 3032,
		message:
			'Any data returned from middleware must be a valid `Response` object.',
	},
```
Would this be accurate to say? I feel like we don't have to say that NOT doing the thing you have to do will cause errors. I think we can just say, "You have to do this!" and errors are otherwise implied?  :smile: